src/prefect/experimental/workers/base.py

```diff
@@ -0,0 +1,590 @@
+        await self._loop_task_group.__aenter__()
+        await self._runs_task_group.__aenter__()
+        self.is_setup = True
+
+    async def teardown(self, *exc_info):
+        """Cleans up resources after the worker is stopped."""
+        self.logger.debug("Tearing down worker...")
+        self.is_setup = False
+        if self._runs_task_group:
+            self._runs_task_group.cancel_scope.cancel()
+            await self._runs_task_group.__aexit__(*exc_info)
+        if self._loop_task_group:
+            self._loop_task_group.cancel_scope.cancel()
+            await self._loop_task_group.__aexit__(*exc_info)
+        if self._client:
+            await self._client.__aexit__(*exc_info)
+        self._runs_task_group = None
+        self._client = None
+
+    async def _on_start(self):
+        """
+        Starts the heartbeat and storage scan loops when the worker starts.
+        """
+        if not self.is_setup or not self._loop_task_group:
+            raise RuntimeError(
+                f"Worker has not been setup. Use `async with {self.__class__.__name__}()...`"
+            )
+        await super()._on_start()
+        # wait for an initial heartbeat to configure the worker
+        await self.heartbeat_worker()
+        # perform initial scan of storage
+        await self.scan_storage_for_deployments()
+        # schedule the heartbeat loop to run every `heartbeat_seconds`
+        self._loop_task_group.start_soon(
+            partial(
+                critical_service_loop,
+                workload=self.heartbeat_worker,
+                interval=self.heartbeat_seconds,
+                printer=self.logger.debug,
+            )
+        )
+        # schedule the storage scan loop to run
+        self._loop_task_group.start_soon(
+            partial(
+                critical_service_loop,
+                workload=self.scan_storage_for_deployments,
+                interval=self.storage_scan_seconds,
+                printer=self.logger.debug,
+            )
+        )
+
+    async def _on_stop(self) -> None:
+        await super()._on_stop()
+
+    def __repr__(self):
+        return f"Worker(pool={self.worker_pool_name!r}, name={self.name!r})"
+
+    async def run_once(self):
+        # if the pool is paused or has a 0 concurrency limit, don't bother polling
+        if self.worker_pool and (
+            self.worker_pool.is_paused or self.worker_pool.concurrency_limit == 0
+        ):
+            return
+
+        runs_response = await self.get_scheduled_flow_runs()
+        await self._submit_scheduled_flow_runs(flow_run_response=runs_response)
+
+    async def heartbeat_worker(self):
+        """
+        Updates the worker's local information about it's current worker pool and queues.
+        Sends a heart beat to the API.
+        """
+
+        # ----------------------------------------------
+        # Update local worker pool information
+        # ----------------------------------------------
```

Comment:
I feel like if a big header comment like this is useful the contents would probably be better split into a separate helper method.